Do not consume 100% CPU while waiting for upload jobs to complete

mcb30 opened the pull request:
The current logic in SessionContext::UploadLoop() degenerates into a
CPU spinloop while waiting for the final upload jobs to complete.

Fix by using an end-of-queue dummy terminator job instead of an atomic
flag, thereby allowing the loop to sleep instead of spinning.

There seems to be no need to use the atomic compare-and-swap in
FinalizeDerived() to guarantee that pthread_join() is called exactly
once, since the surrounding control flow already appears to require
that FinalizeDerived() is called exactly once.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

Fixes: #2720

mcb30 committed:
Do not consume 100% CPU while waiting for upload jobs to complete

The current logic in SessionContext::UploadLoop() degenerates into a
CPU spinloop while waiting for the final upload jobs to complete.

Fix by using an end-of-queue dummy terminator job instead of an atomic
flag, thereby allowing the loop to sleep instead of spinning.

There seems to be no need to use the atomic compare-and-swap in
FinalizeDerived() to guarantee that pthread_join() is called exactly
once, since the surrounding control flow already appears to require
that FinalizeDerived() is called exactly once.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cloudtest

jblomer commented on cvmfs/session_context.cc:
```diff
@@ -299,8 +294,8 @@ bool SessionContext::InitializeDerived(uint64_t max_queue_size) {
 bool SessionContext::FinalizeDerived() {
-  if (atomic_cas32(&worker_terminate_, 0, 1))
-    pthread_join(worker_, NULL);
+  upload_jobs_->Enqueue(&terminator_);
```
I think we should add a comment here.  If I'm not mistaken, `Spooler::Construct` will ensure that `SessionContext::Initialize()` is called so that the worker thread is present when `Spooler::FinalizeSession()` calls `GatewayUploader::FinalizeSession()` calls `SessionContext::Finalize()`. The comment should explain why there has to be a worker thread.  I think we should also add a `TODO` to add a factory method to the `SessionContext` class so that we cannot construct it without also initializing it.

mcb30 replied:
I'm happy to add a comment, but I don't think I understand the control flow here well enough to document why the worker thread needs to exist (which it already did prior to this commit).  What should the comment say?

jblomer commented:
@cernvm-bot cloudtest

jblomer commented on cvmfs/session_context.cc:
```diff
@@ -409,30 +404,27 @@ bool SessionContext::DoUpload(const SessionContext::UploadJob* job) {
-      jobs_processed++;
     }
-    if (ctx->queue_was_flushed_.IsEmpty()) {
+    if (ctx->upload_jobs_->IsEmpty() && ctx->queue_was_flushed_.IsEmpty()) {
```
The logic `if (IsEmpty()) Enqueue();` smells like a race condition. I realize that it was there before, I need another moment to think it through.

mcb30 replied:
Agreed that the code needs more fixing.  I don't think this PR adds any race conditions that weren't already present, but it does at least remove a CPU spinloop.  Could fixing the race condition(s) be deferred for a later fix?

jblomer replied:
Boy, these classes need a rewrite.

If I'm not mistaken, we can get rid of `queue_was_flushed_` entirely.  It seems to me not to serve any useful purpose.  The way it is implemented, it blocks the `WaitForUpload()` call until a submitted object pack in flight is sent and submitted to the gateway.  However, that logic only starts as of the second `WaitForUpload()` call because the `queue_was_flushed_` is initially `true`.  Also, the upload queue works on entire object packs -- so as long as an object pack is not full (and hence not dispatched), `WaitForUpload()` doesn't wait.

The barrier for the gateway uploader is actually not `WaitForUpload()` but `FinalizeSession()`.  Finalizing the session will end up in `SessionContextBase::Finalize`, which looks at the currently open object pack and closes out the tab itself.

Let's try (in this PR) removing `queue_was_flushed_` and just returning from `WaitForUpload()`.

mcb30 replied:
Added an extra commit to remove `queue_was_flushed_` completely, as described.  I put you as the commit author since my contribution here is merely secretarial.

mcb30 replied:
As a quick smoke test: running tests 800 and 802 still passes after making this change.

jblomer commented on cvmfs/session_context.cc:
```diff
@@ -299,8 +294,8 @@ bool SessionContext::InitializeDerived(uint64_t max_queue_size) {
 bool SessionContext::FinalizeDerived() {
-  if (atomic_cas32(&worker_terminate_, 0, 1))
-    pthread_join(worker_, NULL);
+  upload_jobs_->Enqueue(&terminator_);
```
```suggestion
  // Note: in FinalizedDerived, we know that the worker is running. The SessionContext is only called from
  // GatewayUploader::FinalizeSession(), which in turn is from Spooler::FinalizeSession().  The Spooler ensures that
  // GatewayUploader::Initialize() is called on construction.
  // TODO(jblomer): Refactor SessionContext (and Uploader*) classes to use a factory method for construction.
  upload_jobs_->Enqueue(&terminator_);
```

mcb30 replied:
Thank you very much - have added this to the commit

jblomer committed:
Remove SessionContextBase::queue_was_flushed_

The queue_was_flushed_ FIFO seems to serve no useful purpose.  The way
it is implemented, it blocks the WaitForUpload() call until a
submitted object pack in flight is sent and submitted to the gateway.
However, that logic only starts as of the second WaitForUpload() call
because the queue_was_flushed_ is initially true.  Also, the upload
queue works on entire object packs - so as long as an object pack is
not full (and hence not dispatched), WaitForUpload() doesn't wait.

The barrier for the gateway uploader is actually not WaitForUpload()
but FinalizeSession().  Finalizing the session will end up in
SessionContextBase::Finalize, which looks at the currently open object
pack and closes out the tab itself.

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot cpplint

mcb30 commented:
Are there any more changes needed?  It looks as though the single cloudtest failure was unrelated.
